conda/gateways/repodata/__init__.py

```diff
@@ -289,3 +317,434 @@ def conda_http_errors(url, repodata_fn):
+
+        try:
+            obj = self[key]
+            last_checked_str = obj["last_checked"]
+            if last_checked_str.endswith("Z"):
+                last_checked_str = f"{last_checked_str[:-1]}+00:00"
+            last_checked = datetime.datetime.fromisoformat(last_checked_str)
+            value = bool(obj["value"])
+            return (value, last_checked)
+        except (KeyError, ValueError, TypeError) as e:
+            log.warn("error parsing `has_` object from `<cache key>.state.json`", exc_info=e)
+            self.pop(key)
+
+        return False, datetime.datetime.now(tz=datetime.timezone.utc)
+
+    def set_has_format(self, format: str, value: bool):
+        key = f"has_{format}"
+        self[key] = {
+            "last_checked": datetime.datetime.now(tz=datetime.timezone.utc).isoformat()[
+                : -len("+00:00")
+            ]
+            + "Z",
+            "value": value,
+        }
+
+    def should_check_format(self, format: str) -> bool:
+        """
+        Return True if named format should be attempted.
+        """
+        has, when = self.has_format(format)
+        return (
+            has is True
+            or isinstance(when, datetime.datetime)
+            and datetime.datetime.now(tz=datetime.timezone.utc) - when
+            > CHECK_ALTERNATIVE_FORMAT_INTERVAL
+        )
+
+    def __setitem__(self, key: str, item: Any) -> None:
+        if key in self._aliased:
+            key = key[1:]  # strip underscore
+        return super().__setitem__(key, item)
+
+    def __missing__(self, key: str):
+        if key in self._aliased:
+            key = key[1:]  # strip underscore
+        else:
+            raise KeyError(key)
+        return super().__getitem__(key)
+
+
+LOCK_BYTE = 21  # mamba interop
+LOCK_ATTEMPTS = 10
+LOCK_SLEEP = 1
+
+
+@contextmanager
+def _lock_noop(fd):
+    """
+    When locking is not available.
+    """
+    yield
+
+
+try:  # pragma: no cover
+    import msvcrt
+
+    @contextmanager
+    def _lock_impl(fd):  # type: ignore
+        tell = fd.tell()
+        fd.seek(LOCK_BYTE)
+        msvcrt.locking(fd.fileno(), msvcrt.LK_LOCK, 1)  # type: ignore
+        try:
+            fd.seek(tell)
+            yield
+        finally:
+            fd.seek(LOCK_BYTE)
+            msvcrt.locking(fd.fileno(), msvcrt.LK_UNLCK, 1)  # type: ignore
+
+except ImportError:
+    try:
+        import fcntl
+    except ImportError:  # pragma: no cover
+        # "fcntl Availibility: not Emscripten, not WASI."
+        warnings.warn("file locking not available")
+
+        _lock_impl = _lock_noop  # type: ignore
+
+    else:
+
+        class _lock_impl:
+            def __init__(self, fd):
+                self.fd = fd
+
+            def __enter__(self):
+                for attempt in range(LOCK_ATTEMPTS):
+                    try:
+                        # msvcrt locking does something similar
+                        fcntl.lockf(self.fd, fcntl.LOCK_EX | fcntl.LOCK_NB, 1, LOCK_BYTE)
```

Comment:
Don't we need to return here if the locking is successful?

Reply:
Yes. It must complete the other attempts really quickly if the lock is already acquired!

Reply:
It breaks out of the loop now.